File: app/src/main/java/io/github/shadow578/music_dl/db/DBTypeConverters.java
package io.github.shadow578.music_dl.db;

import androidx.room.TypeConverter;

import java.time.LocalDate;

import io.github.shadow578.music_dl.KtPorted;
import io.github.shadow578.music_dl.db.model.TrackStatus;
import io.github.shadow578.music_dl.util.storage.StorageKey;

/**
 * type converters for room
 */
@KtPorted
class DBTypeConverters {

    //region StorageKey
    @TypeConverter
    public static String fromStorageKey(StorageKey key) {
        if (key == null) {
            return null;
        }

        return key.toString();
    }

    @TypeConverter
    public static StorageKey toStorageKey(String key) {
        if (key == null) {
            return null;
        }

        return new StorageKey(key);
    }
    //endregion

    //region TrackStatus
    @TypeConverter
    public static String fromTrackStatus(TrackStatus status) {
        if (status == null) {
            return null;
        }

        return status.key();
    }

    @TypeConverter
    public static TrackStatus toTrackStatus(String key) {
        if (key == null) {
            return null;
        }

        final TrackStatus s = TrackStatus.findByKey(key);
        if (s != null) {
            return s;
        }

        return TrackStatus.DownloadPending;
    }
    //endregion

    //region LocalDate
    @TypeConverter
    public static String fromLocalDate(LocalDate date) {
        if (date == null) {
            return null;
        }

        return date.toString();
    }

    @TypeConverter
    public static LocalDate toLocalDate(String string) {
        if (string == null) {
            return null;
        }

        return LocalDate.parse(string);
    }
    //endregion
}

File: app/src/main/kotlin/io/github/shadow578/yodel/db/DBTypeConverters.kt
package io.github.shadow578.yodel.db

import androidx.room.TypeConverter
import io.github.shadow578.yodel.db.model.TrackStatus
import io.github.shadow578.yodel.util.storage.StorageKey
import java.time.LocalDate

/**
 * type converters for room
 */
class DBTypeConverters {
    //region StorageKey
    @TypeConverter
    fun fromStorageKey(key: StorageKey?): String? {
        return key?.toString()
    }

    @TypeConverter
    fun toStorageKey(key: String?): StorageKey? {
        return if (key == null) {
            null
        } else StorageKey(key)
    }

    //endregion

    //region TrackStatus
    @TypeConverter
    fun fromTrackStatus(status: TrackStatus?): String? {
        return status?.key
    }

    @TypeConverter
    fun toTrackStatus(key: String?): TrackStatus? {
        if (key == null) {
            return null
        }
        val s = TrackStatus.findByKey(key)
        return s ?: TrackStatus.DownloadPending
    }

    //endregion

    //region LocalDate
    @TypeConverter
    fun fromLocalDate(date: LocalDate?): String? {
        return date?.toString()
    }

    @TypeConverter
    fun toLocalDate(string: String?): LocalDate? {
        return if (string == null) {
            null
        } else LocalDate.parse(string)
    } //endregion
}